### PR TITLE
API: Close the underlying iterator in ClosingIterator correctly.

### PR DESCRIPTION
**Pull request by saurabhagas**

### CODE REVIEW

**Comment by vvellanki** [comment]
@rymurr Can you review this patch?

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (openinx) [inline api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:37]: I think you mean in the read pattern (for the original `ClosingIterator`)

```java
   while (hasNext()) {
      T val = next();
    }
```

When the hasNext() return false, then we won't call the `next()` to close the internal `CloseableIterator`.   Yes, I think you are correct,  but I think we will need to guarantee that we will only close the `CloseableIterator` once even if we called the `ClosingIterator#hasNext` many times, otherwise once we've called the first `hasNext`, all the following `hasNext` check will be thrown an exception, that's not the expected behavior.

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:47]: you probably should make sure that `close()` is only called once

Comment (saurabhagas) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:47]: > you probably should make sure that `close()` is only called once

`Closeable::close` is required to be idempotent - if calling close on a `Closeable` resource has any side-effects, we should fix that at the source.

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (saurabhagas) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:37]: `Closeable::close` is required to be idempotent - if calling close on a `Closeable` resource has any side-effects, we should fix that at the source. This would also mean that calling close() multiple times should not throw any exceptions.

Comment (openinx) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:37]: > we should fix that at the source

We can not assume that all the source could guarantee it only close one times, even if all the resources introduced in apache iceberg project don't guarantee it.  The correct approach is keeping it close once in this `ClosingIterator`. I think introducing a boolean `closed` is enough ?

Comment (rymurr) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:37]: > We can not assume that all the source could guarantee it only close one times, even if all the resources introduced in apache iceberg project don't guarantee it.

I agree

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (rymurr) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:47]: A cursory search of various closable iterators in the iceberg project indicates that not all are idempotent. I think we should raise a ticket to address the issue in the iceberg project at least. But in the meantime we can just check if this has already been closed.

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (rymurr) [inline api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:26]: @rdblue any reason this is not `Closeable`? The only way for the underlying resource to be closed right now is for this iterator to be exhausted. Should this be `Closable` so the resources can be closed w/o having to read the whole iterator? Or is this by design and we should add javadoc to document this contract?

**Review by rymurr — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (saurabhagas) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:38]: Done, and added a test.

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (saurabhagas) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:48]: Done, and added a test.

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (saurabhagas) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:26]: I think this is by design. The intent must be to avoid leaks because the developer forgot to call `close()` on the iterator.

Comment (rymurr) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:26]: What about the case when the developer doesn't consume the entire iterator but has scheduled this to be closed?

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (rymurr) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:48]: What is the behaviour if htis iterator is being used in parallel?

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (saurabhagas) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:30]: If a premature close is required, the underlying `CloseableIterator` should be used directly. I see this class as a convenience wrapper for the case when all of the elements are intended to be consumed. I've added Javadoc mentioning this.

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (saurabhagas) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:52]: The iterator isn't required to be thread-safe from what I can tell.

**Review by rymurr — APPROVED** [review]
Thanks @saurabhagas will merge when the build fiinishes

**Comment by saurabhagas** [comment]
Thanks @rymurr and other reviewers for accepting the change.

**Review by nastra — APPROVED** [review]
LGTM

**Comment by openinx** [comment]
@rymurr , I think we will need to check whether the contributor are following the commit message pattern: `[iceberg-module]: [commit log message]`.  For example, this commit message should be rewrite as `API: Close the underlying iterator in ClosingIterator correctly`.  If the contributor did not follow the pattern, we committer will need to change it to (I think it's also good to add this to the apache iceberg [document](https://iceberg.apache.org/community/#contributing)).  We don't need to revert this committed PR, please have an extra check next time :-) .

**Comment by rymurr** [comment]
ahh, thanks for the check @openinx you are absolutely right. I will be more cautious next time.

File: api/src/main/java/org/apache/iceberg/io/ClosingIterator.java
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/ClosingIterator.java:30]: I think this was an oversight. I don't see a reason why it shouldn't be `Closeable`.

**Pull request merged**
